machine-learning/solutions/seismic_salt_detection/trainer/inputs.py

```diff
@@ -107,33 +131,48 @@ def _input_fn():
         return features, target
     return _input_fn
 
+def get_serving_function(image_path):
+    """Creates a serving function.
 
-def csv_serving_input_fn():
-    """Creates a `ServingInputReceiver` for inference.
-
-    Creates a  placeholder for the record and specifies how to parse the
-    features from that record.
+    Args:
+        image_path: Path to image folder.
 
     Returns:
-        A `ServingInputReceiver`.
+        Serving function.
     """
-    csv_row = tf.placeholder(
-        dtype=tf.string
-    )
+    def csv_serving_input_fn():
+        """Creates a `ServingInputReceiver` for inference.
+
+        Creates a  placeholder for the record and specifies how to parse the
+        features from that record.
+
+        Returns:
+            A `ServingInputReceiver`.
+        """
+        csv_row = tf.placeholder(
+            shape=[None],
+            dtype=tf.string
+        )
 
-    features = parse_csv(csv_row)
-    features, _ = get_features_target_tuple(features)
+        features = parse_csv(csv_row)
+        features, _ = get_features_target_tuple(features)
+        features['image'] = tf.map_fn(
```

Comment:
I am highly inexpert when it comes to TensorFlow.  Why did you choose map_fn for what looks like simple function application?

Reply:
map_fn helps you apply the function in parallel (and get the output in the desired shape).
This is equivalent to the dataset.map I used for training. tf.map_fn helps me do the same at inference for several records at the same time.

Reply:
I think of a dataset as like a collection of records.  Dataset.map(f) maps f over each record.

tf.map_fn(f, c) similarly maps a function over a collection.  And as per your comment, I think I understand how it helps TF apply the function in parallel.

But here you're using it even when collection contains just a single item.  Is there parallelism that I'm not seeing?

Reply:
The size of the tensor is not known:
csv_row = tf.placeholder(
            shape=[None],
            dtype=tf.string
        )
Note shape [None]^

That means that when we publish this model people can call it with how many records as they want (or fit in memory) and we'll process all in parallel (CMLE batch prediction product always calls this function with 50 examples every time. I think they hardcoded that number on their side)

Reply:
features['id']  is a tensor collection of strings at that level, but if you do tf.map_fn(features['id']) (or dataset.map()) it applies a specific function to each element of that collection (separately but in parallel) and returns a collection with all elements transformed by that function.